01-architecture.md

```diff
@@ -0,0 +1,383 @@
+
+Admittedly, this is an inefficient solution in terms of performance, but it allows for straightforward and relatively risk-free deployment of existing battle-tested Ethereum smart contracts to the Filecoin network.
+
+The gas accounting will factor in the inefficiency, resulting in more expensive executions. This will incentivise developers to migrate the smart contracts to native FVM actors to attain lower execution costs.
+
+In addition to the EVM, in the future we are keen to support for Agoric SES, Solana's BPF, and other blockchain programming models and paradigms.
+
+We believe that compatibility should be accomplished by translating/emulating the lowest-possible executable output in its source form, rather than dealing with high-level languages using alternative/custom toolchains.
+
+Moreover, this choice enables developers to (re-)use all the tooling available in the source ecosystems, and results in the highest possible execution fidelity/parity, thus reducing technical risks.
+
+Refer to [foreign runtimes](#foreign-runtimes) for more detail.
+
+### Built-in system actors
+
+The existing built-in system actors will be compiled to WASM, and will be made to use the FVM SDK. They will run entirely in WASM space.
+
+For the sake of analogy, built-in system actors in Filecoin receive comparable treatment to "precompiled contracts" in other platforms. They are specified by the protocol, implicitly deployed, addressed at fixed locations/mailboxes, and they evolve through system upgrades.
+
+**Canonical system actors**
+
+This transition to WASM opens up the opportunity to converge on a single codebase of actors for all Filecoin implementations.
+
+In the past, each team would re-implement actors in their language (although some relied on FFI with Go actors, a slower approach). With FVM, a single codebase can be compiled to WASM (its new portable executable form) and be adopted across all implementations.
+
+We acknowledge this strategy has tradeoffs, but we won't elaborate on them here.
+
+## Execution architecture
+
+![FVM execution architecture](img/fvm-exec-arch.png)
+
+### FVM
+
+The FVM is implemented in Rust. It is responsible for:
+
+1. Setting up Invocation Containers, managing their lifetime and stacking. This include setting up the WASM runtime, its managed memory, binding syscalls as host-provided functions, and potentially mapping dynamically-linked libraries (e.g. predefined sdk versions).
+2. Serving as a call orchestration layer across actors.
+3. Resolving Boundary B syscalls where possible within the FVM.
+4. Relaying Boundary B syscalls to Boundary A syscalls where the above is unfeasible.
+5. Managing IPLD state tree data, by buffering writes from actors running in ICs until execution succeeds, maintaining caches, optimising access patterns (e.g. optimistic fetches), and more.
+
+### Invocation Container (IC)
```

Comment:
> 3. We have the "context" related to the invocation (message, sender, receiver, reachable IPLD data and IPLD caches, etc.). This is the component that will provide per-invocation syscall implementations. <-- the missing thing?

In the current model, the thing that tracks the context of every call is the FVM, including the IPLD data/caches, etc. The call stack manager inside the FVM initialises the context bound to each invocation container and links/binds the injected syscalls to the context. That's how I imagined it working in my head, but admittedly this could work differently, but I'm not sure how much would be an implementation detail, and how much would be mandated by the architecture.